File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        // 2 pointer approach, TC => O(n)
        int mostWater = Integer.MIN_VALUE;
        int lp = 0, rp = height.length - 1;
        while(lp < rp) {
            int ht = Math.min(height[lp], height[rp]);
            int wdth = rp - lp;
            mostWater = Math.max(mostWater, (ht*wdth));
            if(height[lp] < height[rp]) {
                lp++;
            }
            else{
                rp--;
            }
        }
        return mostWater;
    }
}

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int mostWater = INT_MIN;
        int i=0, j = height.size() - 1;
        while(i < j){
            int ht = min(height[i], height[j]);
            mostWater = max(mostWater, ht*(j - i));
            if(height[i] < height[j]) i++;
            else j--;
        }
        return mostWater;
    }
};